src/lib/textToEdit.ts

```diff
@@ -0,0 +1,351 @@
+
+        while (timeElapsed < MAX_CHECK_TIMEOUT) {
+          const check = await getPromptToEditResult(submitResult.id, token)
+          console.log('check', check)
+          if (check instanceof Error || check.status === 'failed') {
+            console.log('check failed', check)
+            reject(check)
+            return
+          } else if (check.status === 'completed') {
+            resolve(check)
+            return
+          }
+
+          await new Promise((r) => setTimeout(r, CHECK_DELAY))
+          timeElapsed += CHECK_DELAY
+        }
+
+        reject(new Error('Text-to-CAD API timed out'))
+      })().catch(reportRejection)
+    }
+  )
+
+  try {
+    const result = await textToCadComplete
+    toast.dismiss(toastId)
+    console.log('textToCadComplete', result)
+    return result
+  } catch (e) {
+    toast.dismiss(toastId)
+    toast.error(
+      'Failed to edit your KCL code, please try again with a different prompt or selection'
+    )
+    console.error('textToCadComplete', e)
+  }
+
+  return textToCadComplete
+}
+
+export async function doWholeFlow({
+  prompt,
+  selections,
+  code,
+  token,
+  artifactGraph,
+}: {
+  prompt: string
+  selections: Selections
+  code: string
+  token?: string
+  artifactGraph: ArtifactGraph
+}) {
+  const result = await doPromptEdit({
+    prompt,
+    selections,
+    code,
+    token,
+    artifactGraph,
+  })
+  if (err(result)) return Promise.reject(result)
+  console.log('result', result)
+  const oldCode = codeManager.code
+  const { code: newCode } = result
+  codeManager.updateCodeEditor(newCode)
+  const diff = reBuildNewCodeWithRanges(oldCode, newCode)
+  const ranges: SelectionRange[] = diff.insertRanges.map((range) =>
+    EditorSelection.range(range[0], range[1])
+  )
+  editorManager?.editorView?.dispatch({
+    selection: EditorSelection.create(
+      ranges,
+      selections.graphSelections.length - 1
+    ),
+    annotations: [modelingMachineEvent, Transaction.addToHistory.of(false)],
+  })
+  await kclManager.executeCode()
+  const toastId = uuidv4()
+
+  toast.success(
+    () =>
+      ToastPromptToEditCadSuccess({
+        toastId,
+        data: result,
+        token,
+        oldCode,
+      }),
+    {
+      id: toastId,
+      duration: Infinity,
+      icon: null,
+    }
+  )
+}
+
+const reBuildNewCodeWithRanges = (
+  oldCode: string,
+  newCode: string
+): {
+  newCode: string
+  insertRanges: SourceRange[]
+} => {
+  let insertRanges: SourceRange[] = []
+  const changes = diffLines(oldCode, newCode)
+  let newCodeWithRanges = ''
+  for (const change of changes) {
+    if (!change.added && !change.removed) {
+      // no change add it to newCodeWithRanges
+      newCodeWithRanges += change.value
+    } else if (change.added && !change.removed) {
+      const start = newCodeWithRanges.length
+      const end = start + change.value.length
+      insertRanges.push([start, end, true])
+      newCodeWithRanges += change.value
+    }
```

Comment:
I can't think of a good way to show the user that lines were deleted 🤔

Reply:
Yeah not without an actual diff UI.